source/client_channel_handler.c

```diff
@@ -728,53 +762,57 @@ static void s_request_timeout_task(struct aws_channel_task *task, void *arg, enu
                 "id=%p: removing message id %" PRIu16 " from the outstanding requests list.",
-                (void *)request->connection,
-                request->message_id);
-            aws_mutex_lock(&request->connection->outstanding_requests.mutex);
-
-            aws_hash_table_remove(
-                &request->connection->outstanding_requests.table, &request->message_id, &elem, &was_present);
-
-            aws_mutex_unlock(&request->connection->outstanding_requests.mutex);
-
-            AWS_ASSERT(was_present);
+                (void *)connection,
+                request->packet_id);
 
             AWS_LOGF_TRACE(
                 AWS_LS_MQTT_CLIENT,
                 "id=%p: (timeout task run of now-completed request) Releasing request %" PRIu16
                 " to connection memory pool",
-                (void *)(request->connection),
-                request->message_id);
-
-            struct aws_mqtt_client_connection *connection = request->connection;
-            aws_mutex_lock(&connection->requests_pool.mutex);
-            aws_memory_pool_release(&connection->requests_pool.pool, elem.value);
-            aws_mutex_unlock(&connection->requests_pool.mutex);
-        } else if (request->connection->state == AWS_MQTT_CLIENT_STATE_CONNECTED) {
-            /* If not complete and online, schedule retry task */
-
-            uint64_t ttr = 0;
-            aws_channel_current_clock_time(request->connection->slot->channel, &ttr);
-            ttr += request->connection->request_timeout_ns;
-            AWS_LOGF_TRACE(
-                AWS_LS_MQTT_CLIENT,
-                "id=%p: scheduling timeout task for message id %" PRIu16 " to run at %" PRIu64,
-                (void *)request->connection,
-                request->message_id,
-                ttr);
-
-            aws_channel_schedule_task_future(request->connection->slot->channel, task, ttr);
-        } else {
-            /* Else, put the task in the pending list */
-            AWS_LOGF_TRACE(
-                AWS_LS_MQTT_CLIENT,
-                "id=%p: not currently connected, adding message id %" PRIu16 " to the pending requests list.",
-                (void *)request->connection,
-                request->message_id);
-            aws_mutex_lock(&request->connection->pending_requests.mutex);
-            aws_linked_list_push_back(&request->connection->pending_requests.list, &request->list_node);
-            aws_mutex_unlock(&request->connection->pending_requests.mutex);
+                (void *)(connection),
+                request->packet_id);
+            { /* BEGIN CRITICAL SECTION */
+                s_mqtt_connection_lock_synced_data(connection);
+                aws_hash_table_remove(
+                    &connection->synced_data.outstanding_requests_table, &request->packet_id, &elem, &was_present);
+                /* If the request is not in outstanding_requests_table, that's a bug */
+                AWS_ASSERT(was_present);
```

Comment:
This should be a warning. It shouldn't exit the program, since the state is already what we wanted: the table doesn't contain the request